File: ethereumj-core/src/test/java/org/ethereum/jsontestsuite/suite/validators/BlockHeaderValidator.java
/*
 * The MIT License (MIT)
 *
 * Copyright 2017 Alexander Orlov <alexander.orlov@loxal.net>. All rights reserved.
 * Copyright (c) [2016] [ <ether.camp> ]
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 *
 */

package org.ethereum.jsontestsuite.suite.validators;

import org.ethereum.core.BlockHeader;

import java.math.BigInteger;
import java.util.ArrayList;

import static org.ethereum.util.ByteUtil.toHexString;

public class BlockHeaderValidator {


    public static ArrayList<String> valid(final BlockHeader orig, final BlockHeader valid) {

        final ArrayList<String> outputSummary = new ArrayList<>();

        if (!toHexString(orig.getParentHash())
                .equals(toHexString(valid.getParentHash()))) {

            final String output =
                    String.format("wrong block.parentHash: \n expected: %s \n got: %s",
                            toHexString(valid.getParentHash()),
                            toHexString(orig.getParentHash())
                    );

            outputSummary.add(output);
        }

        if (!toHexString(orig.getUnclesHash())
                .equals(toHexString(valid.getUnclesHash()))) {

            final String output =
                    String.format("wrong block.unclesHash: \n expected: %s \n got: %s",
                            toHexString(valid.getUnclesHash()),
                            toHexString(orig.getUnclesHash())
                    );

            outputSummary.add(output);
        }

        if (!toHexString(orig.getCoinbase())
                .equals(toHexString(valid.getCoinbase()))) {

            final String output =
                    String.format("wrong block.coinbase: \n expected: %s \n got: %s",
                            toHexString(valid.getCoinbase()),
                            toHexString(orig.getCoinbase())
                    );

            outputSummary.add(output);
        }

        if (!toHexString(orig.getStateRoot())
                .equals(toHexString(valid.getStateRoot()))) {

            final String output =
                    String.format("wrong block.stateRoot: \n expected: %s \n got: %s",
                            toHexString(valid.getStateRoot()),
                            toHexString(orig.getStateRoot())
                    );

            outputSummary.add(output);
        }

        if (!toHexString(orig.getTxTrieRoot())
                .equals(toHexString(valid.getTxTrieRoot()))) {

            final String output =
                    String.format("wrong block.txTrieRoot: \n expected: %s \n got: %s",
                            toHexString(valid.getTxTrieRoot()),
                            toHexString(orig.getTxTrieRoot())
                    );

            outputSummary.add(output);
        }

        if (!toHexString(orig.getReceiptsRoot())
                .equals(toHexString(valid.getReceiptsRoot()))) {

            final String output =
                    String.format("wrong block.receiptsRoot: \n expected: %s \n got: %s",
                            toHexString(valid.getReceiptsRoot()),
                            toHexString(orig.getReceiptsRoot())
                    );

            outputSummary.add(output);
        }

        if (!toHexString(orig.getLogsBloom())
                .equals(toHexString(valid.getLogsBloom()))) {

            final String output =
                    String.format("wrong block.logsBloom: \n expected: %s \n got: %s",
                            toHexString(valid.getLogsBloom()),
                            toHexString(orig.getLogsBloom())
                    );

            outputSummary.add(output);
        }

        if (!toHexString(orig.getDifficulty())
                .equals(toHexString(valid.getDifficulty()))) {

            final String output =
                    String.format("wrong block.difficulty: \n expected: %s \n got: %s",
                            toHexString(valid.getDifficulty()),
                            toHexString(orig.getDifficulty())
                    );

            outputSummary.add(output);
        }

        if (orig.getTimestamp() != valid.getTimestamp()) {

            final String output =
                    String.format("wrong block.timestamp: \n expected: %d \n got: %d",
                            valid.getTimestamp(),
                            orig.getTimestamp()
                    );

            outputSummary.add(output);
        }

        if (orig.getNumber() != valid.getNumber()) {

            final String output =
                    String.format("wrong block.number: \n expected: %d \n got: %d",
                            valid.getNumber(),
                            orig.getNumber()
                    );

            outputSummary.add(output);
        }

        if (!new BigInteger(1, orig.getGasLimit()).equals(new BigInteger(1, valid.getGasLimit()))) {

            final String output =
                    String.format("wrong block.gasLimit: \n expected: %d \n got: %d",
                            new BigInteger(1, valid.getGasLimit()),
                            new BigInteger(1, orig.getGasLimit())
                    );

            outputSummary.add(output);
        }

        if (orig.getGasUsed() != valid.getGasUsed()) {

            final String output =
                    String.format("wrong block.gasUsed: \n expected: %d \n got: %d",
                            valid.getGasUsed(),
                            orig.getGasUsed()
                    );

            outputSummary.add(output);
        }

        if (!toHexString(orig.getMixHash())
                .equals(toHexString(valid.getMixHash()))) {

            final String output =
                    String.format("wrong block.mixHash: \n expected: %s \n got: %s",
                            toHexString(valid.getMixHash()),
                            toHexString(orig.getMixHash())
                    );

            outputSummary.add(output);
        }

        if (!toHexString(orig.getExtraData())
                .equals(toHexString(valid.getExtraData()))) {

            final String output =
                    String.format("wrong block.extraData: \n expected: %s \n got: %s",
                            toHexString(valid.getExtraData()),
                            toHexString(orig.getExtraData())
                    );

            outputSummary.add(output);
        }

        if (!toHexString(orig.getNonce())
                .equals(toHexString(valid.getNonce()))) {

            final String output =
                    String.format("wrong block.nonce: \n expected: %s \n got: %s",
                            toHexString(valid.getNonce()),
                            toHexString(orig.getNonce())
                    );

            outputSummary.add(output);
        }


        return outputSummary;
    }
}

File: free-ethereum-core/src/test/java/org/ethereum/jsontestsuite/suite/validators/BlockHeaderValidator.kt
/*
 * The MIT License (MIT)
 *
 * Copyright 2017 Alexander Orlov <alexander.orlov@loxal.net>. All rights reserved.
 * Copyright (c) [2016] [ <ether.camp> ]
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 *
 */

package org.ethereum.jsontestsuite.suite.validators

import org.ethereum.core.BlockHeader

import java.math.BigInteger
import java.util.ArrayList

import org.ethereum.util.ByteUtil.toHexString

object BlockHeaderValidator {


    fun valid(orig: BlockHeader, valid: BlockHeader): ArrayList<String> {

        val outputSummary = ArrayList<String>()

        if (toHexString(orig.parentHash) != toHexString(valid.parentHash)) {

            val output = String.format("wrong block.parentHash: \n expected: %s \n got: %s",
                    toHexString(valid.parentHash),
                    toHexString(orig.parentHash)
            )

            outputSummary.add(output)
        }

        if (toHexString(orig.unclesHash) != toHexString(valid.unclesHash)) {

            val output = String.format("wrong block.unclesHash: \n expected: %s \n got: %s",
                    toHexString(valid.unclesHash),
                    toHexString(orig.unclesHash)
            )

            outputSummary.add(output)
        }

        if (toHexString(orig.coinbase) != toHexString(valid.coinbase)) {

            val output = String.format("wrong block.coinbase: \n expected: %s \n got: %s",
                    toHexString(valid.coinbase),
                    toHexString(orig.coinbase)
            )

            outputSummary.add(output)
        }

        if (toHexString(orig.stateRoot) != toHexString(valid.stateRoot)) {

            val output = String.format("wrong block.stateRoot: \n expected: %s \n got: %s",
                    toHexString(valid.stateRoot),
                    toHexString(orig.stateRoot)
            )

            outputSummary.add(output)
        }

        if (toHexString(orig.txTrieRoot) != toHexString(valid.txTrieRoot)) {

            val output = String.format("wrong block.txTrieRoot: \n expected: %s \n got: %s",
                    toHexString(valid.txTrieRoot),
                    toHexString(orig.txTrieRoot)
            )

            outputSummary.add(output)
        }

        if (toHexString(orig.receiptsRoot) != toHexString(valid.receiptsRoot)) {

            val output = String.format("wrong block.receiptsRoot: \n expected: %s \n got: %s",
                    toHexString(valid.receiptsRoot),
                    toHexString(orig.receiptsRoot)
            )

            outputSummary.add(output)
        }

        if (toHexString(orig.logsBloom) != toHexString(valid.logsBloom)) {

            val output = String.format("wrong block.logsBloom: \n expected: %s \n got: %s",
                    toHexString(valid.logsBloom),
                    toHexString(orig.logsBloom)
            )

            outputSummary.add(output)
        }

        if (toHexString(orig.difficulty) != toHexString(valid.difficulty)) {

            val output = String.format("wrong block.difficulty: \n expected: %s \n got: %s",
                    toHexString(valid.difficulty),
                    toHexString(orig.difficulty)
            )

            outputSummary.add(output)
        }

        if (orig.timestamp != valid.timestamp) {

            val output = String.format("wrong block.timestamp: \n expected: %d \n got: %d",
                    valid.timestamp,
                    orig.timestamp
            )

            outputSummary.add(output)
        }

        if (orig.number != valid.number) {

            val output = String.format("wrong block.number: \n expected: %d \n got: %d",
                    valid.number,
                    orig.number
            )

            outputSummary.add(output)
        }

        if (BigInteger(1, orig.gasLimit) != BigInteger(1, valid.gasLimit)) {

            val output = String.format("wrong block.gasLimit: \n expected: %d \n got: %d",
                    BigInteger(1, valid.gasLimit),
                    BigInteger(1, orig.gasLimit)
            )

            outputSummary.add(output)
        }

        if (orig.gasUsed != valid.gasUsed) {

            val output = String.format("wrong block.gasUsed: \n expected: %d \n got: %d",
                    valid.gasUsed,
                    orig.gasUsed
            )

            outputSummary.add(output)
        }

        if (toHexString(orig.mixHash) != toHexString(valid.mixHash)) {

            val output = String.format("wrong block.mixHash: \n expected: %s \n got: %s",
                    toHexString(valid.mixHash),
                    toHexString(orig.mixHash)
            )

            outputSummary.add(output)
        }

        if (toHexString(orig.extraData) != toHexString(valid.extraData)) {

            val output = String.format("wrong block.extraData: \n expected: %s \n got: %s",
                    toHexString(valid.extraData),
                    toHexString(orig.extraData)
            )

            outputSummary.add(output)
        }

        if (toHexString(orig.nonce) != toHexString(valid.nonce)) {

            val output = String.format("wrong block.nonce: \n expected: %s \n got: %s",
                    toHexString(valid.nonce),
                    toHexString(orig.nonce)
            )

            outputSummary.add(output)
        }


        return outputSummary
    }
}
